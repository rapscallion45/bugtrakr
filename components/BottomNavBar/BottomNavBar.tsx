import { FC } from 'react';
import BottomNavigation from '@mui/material/BottomNavigation';
import BottomNavigationAction from '@mui/material/BottomNavigationAction';
import Link from '../Link/Link';
import useBottomNavBarController from './BottomNavBarController';
import HideOnTrigger from '../HideOnTrigger/HideOnTrigger';

const BottomNavBar: FC = function BottomNavBar() {
  const { showBottomBar, hideBottomBarTrigger, value, routes, onChange } =
    useBottomNavBarController();

  return showBottomBar ? (
    <HideOnTrigger trigger={!hideBottomBarTrigger} direction="up">
      <BottomNavigation value={value} onChange={onChange} showLabels>
        {routes.map((route) => (
          <BottomNavigationAction
            component={Link}
            key={route.value}
            label={route.label}
            value={route.value}
            icon={<route.icon />}
            href={route.path}
          />
        ))}
      </BottomNavigation>
    </HideOnTrigger>
  ) : null;
};
export default BottomNavBar;
