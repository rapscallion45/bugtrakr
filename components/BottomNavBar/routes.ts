import AppsIcon from '@mui/icons-material/Apps';
import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import BugReportIcon from '@mui/icons-material/BugReport';

const routes = [
  {
    path: '/dashboard/my-bugs',
    label: 'My Bugs',
    value: '/dashboard/my-bugs',
    icon: BugReportIcon,
  },
  {
    path: '/dashboard',
    label: 'All Projects',
    value: '/dashboard',
    icon: AppsIcon,
  },
  {
    path: '/dashboard/my-account',
    label: 'My Account',
    value: '/dashboard/my-account',
    icon: AccountCircleIcon,
  },
];
export default routes;
