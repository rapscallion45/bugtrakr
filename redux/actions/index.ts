export { default as alertActions } from './alert.actions';
export { default as accountActions } from './account.actions';
export { default as projectActions } from './project.actions';
export { default as bugActions } from './bug.actions';
export { default as usersActions } from './users.actions';
