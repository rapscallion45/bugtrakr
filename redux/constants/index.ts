export { default as alertConstants } from './alert.constants';
export { default as accountConstants } from './account.constants';
export { default as projectConstants } from './project.constants';
export { default as bugConstants } from './bug.constants';
export { default as usersConstants } from './users.constants';
