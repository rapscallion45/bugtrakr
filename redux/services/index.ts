export { default as accountService } from './account.service';
export { default as projectService } from './project.service';
export { default as bugService } from './bug.service';
export { default as noteService } from './note.service';
export { default as usersService } from './users.service';
