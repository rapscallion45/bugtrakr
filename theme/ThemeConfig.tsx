import { FC, useMemo } from 'react';
import { CacheProvider } from '@emotion/react';
import { CssBaseline } from '@mui/material';
import { ThemeProvider, createTheme } from '@mui/material/styles';
import shape from './shape';
import palette from './palette';
import typography from './typography';
import componentsOverride from './overrides';
import shadows, { customShadows } from './shadows';

interface ThemeConfigProps {
  emotionCache?: any;
  children?: any;
}

const ThemeConfig: FC<ThemeConfigProps> = function ThemeConfig({ emotionCache, children }) {
  const themeOptions = useMemo(
    () => ({
      palette,
      shape,
      typography,
      shadows,
      customShadows,
    }),
    []
  );

  // @ts-ignore
  const theme = createTheme(themeOptions);
  theme.components = componentsOverride(theme);

  return (
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </CacheProvider>
  );
};

export default ThemeConfig;
