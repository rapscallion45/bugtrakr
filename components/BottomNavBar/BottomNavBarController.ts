import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { useRouter } from 'next/router';
import useMediaQuery from '@mui/material/useMediaQuery';
import routes from './routes';

const useBottomNavBarController = () => {
  const isDesktop = useMediaQuery('(min-width:992px)');
  const router = useRouter();
  const [value, setValue] = useState(router.pathname);
  const loggedIn = useSelector((state) => state.authentication.loggedIn);

  useEffect(() => {
    setValue(router.pathname);
  }, [router.pathname]);

  const onChange = (event, newValue) => {
    setValue(newValue);
  };

  const showBottomBar = !isDesktop && loggedIn;
  const hideBottomBarTrigger = useMediaQuery('(min-height:500px)');

  return {
    showBottomBar,
    hideBottomBarTrigger,
    value,
    routes,
    onChange,
  };
};
export default useBottomNavBarController;
