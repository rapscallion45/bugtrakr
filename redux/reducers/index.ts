import { combineReducers } from 'redux';

import registration from './registration.reducer';
import verifyEmail from './verifyEmail.reducer';
import changePassword from './changePassword.reducer';
import resetPassword from './resetPassword.reducer';
import validateResetToken from './validateResetToken.reducer';
import account from './account.reducer';
import alert from './alert.reducer';
import projects from './projects.reducer';
import bugs from './bugs.reducer';
import users from './users.reducer';

const rootReducer = combineReducers({
  account,
  registration,
  verifyEmail,
  changePassword,
  resetPassword,
  validateResetToken,
  alert,
  projects,
  bugs,
  users,
});
export default rootReducer;

export type AppState = ReturnType<typeof rootReducer>;
