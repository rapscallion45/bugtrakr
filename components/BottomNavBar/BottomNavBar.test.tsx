import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import createEmotionCache from '../../utils/createEmotionCache';
import ThemeConfig from '../../theme/ThemeConfig';
import rootReducer from '../../redux/reducers';
import BottomNavBar from './BottomNavBar';
import routes from './routes';

let mockHideBottomBar = false;
let mockRouteValue = '/';

jest.mock(
  '@mui/material/useMediaQuery',
  () =>
    function mock() {
      return mockHideBottomBar;
    }
);

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: mockRouteValue,
      pathname: mockRouteValue,
      query: '',
      asPath: '',
      push: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

jest.mock(
  'next/link',
  () =>
    ({ children }) =>
      children
);

describe('Bottom Nav Bar', () => {
  it('Renders correctly enzyme', () => {
    /* Arrange */
    const testStore = createStore(rootReducer, { authentication: { loggedIn: true } });

    /* Act */
    const view = render(
      <Provider store={testStore}>
        <ThemeConfig emotionCache={createEmotionCache()}>
          <BottomNavBar />
        </ThemeConfig>
      </Provider>
    );

    /* Assert */
    expect(toJson(view)).toMatchSnapshot();
  });

  describe('Show/Hide Bar', () => {
    it('Should not render if show flag not set', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, { authentication: { loggedIn: true } });
      mockHideBottomBar = true;

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <BottomNavBar />
          </ThemeConfig>
        </Provider>
      );

      /* Assert */
      expect(view.find('ForwardRef(BottomNavigation)').length).toBe(0);
    });

    it('Should not render if user logged out', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, { authentication: { loggedIn: false } });
      mockHideBottomBar = false;

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <BottomNavBar />
          </ThemeConfig>
        </Provider>
      );

      /* Assert */
      expect(view.find('ForwardRef(BottomNavigation)').length).toBe(0);
    });
  });

  describe('Nav Routes', () => {
    it('Should have correct number of passed nav items', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, { authentication: { loggedIn: true } });
      mockHideBottomBar = false;

      /* Act */
      const view = render(
        <Provider store={testStore}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <BottomNavBar />
          </ThemeConfig>
        </Provider>
      );

      /* Assert */
      expect(view.children().length).toBe(routes.length);
    });

    it('Should set nav item to active according to current route', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, { authentication: { loggedIn: true } });
      mockRouteValue = '/articles';

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <BottomNavBar />
          </ThemeConfig>
        </Provider>
      );

      /* Assert */
      expect(view.find('ForwardRef(BottomNavigation)').prop('value')).toBe(mockRouteValue);
    });

    it('Should change current app route when nav item clicked', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, { authentication: { loggedIn: true } });
      let testRoute = '/dashboard/my-bugs';
      mockRouteValue = '/articles';

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <BottomNavBar />
          </ThemeConfig>
        </Provider>
      );

      /* Assert */
      expect(view.find('ForwardRef(BottomNavigation)').prop('value')).toBe(mockRouteValue);

      /* Act */
      view
        .find(`a`)
        .at(routes.map((route) => route.path).indexOf(testRoute))
        .simulate('click');

      /* Assert */
      expect(view.find('ForwardRef(BottomNavigation)').prop('value')).toBe(testRoute);

      /* Arrange */
      testRoute = '/dashboard/my-account';

      /* Act */
      view
        .find(`a`)
        .at(routes.map((route) => route.path).indexOf(testRoute))
        .simulate('click');

      /* Assert */
      expect(view.find('ForwardRef(BottomNavigation)').prop('value')).toBe(testRoute);
    });
  });
});
