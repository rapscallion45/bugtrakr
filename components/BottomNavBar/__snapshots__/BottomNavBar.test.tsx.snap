// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bottom Nav Bar Renders correctly enzyme 1`] = `
<div
  class="MuiBottomNavigation-root css-wpgeqa-MuiBottomNavigation-root"
  style="visibility:hidden"
>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiButtonBase-root MuiBottomNavigationAction-root css-1onwxme-MuiTypography-root-MuiLink-root-MuiButtonBase-root-MuiBottomNavigationAction-root"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="BugReportIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20V8zm-6 8h-4v-2h4v2zm0-4h-4v-2h4v2z"
      />
    </svg>
    <span
      class="MuiBottomNavigationAction-label css-1n23l01-MuiBottomNavigationAction-label"
    >
      My Bugs
    </span>
  </a>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiButtonBase-root MuiBottomNavigationAction-root css-1onwxme-MuiTypography-root-MuiLink-root-MuiButtonBase-root-MuiBottomNavigationAction-root"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="AppsIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M4 8h4V4H4v4zm6 12h4v-4h-4v4zm-6 0h4v-4H4v4zm0-6h4v-4H4v4zm6 0h4v-4h-4v4zm6-10v4h4V4h-4zm-6 4h4V4h-4v4zm6 6h4v-4h-4v4zm0 6h4v-4h-4v4z"
      />
    </svg>
    <span
      class="MuiBottomNavigationAction-label css-1n23l01-MuiBottomNavigationAction-label"
    >
      All Projects
    </span>
  </a>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiButtonBase-root MuiBottomNavigationAction-root css-1onwxme-MuiTypography-root-MuiLink-root-MuiButtonBase-root-MuiBottomNavigationAction-root"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="AccountCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
      />
    </svg>
    <span
      class="MuiBottomNavigationAction-label css-1n23l01-MuiBottomNavigationAction-label"
    >
      My Account
    </span>
  </a>
</div>
`;
